feat: implement `sendAndAwaitStatus` subscription (#3541)
FuelLabs/fuels-ts: apps/docs/src/guide/transactions/transaction-subscriptions.md

# Transaction Subscriptions

When submitting transactions via the SDK, usually this is done by calling a method in a [script](../scripts/running-scripts.md) or [contract](../contracts/methods.md#call), or sending a transaction via a [predicate](../predicates/methods.md#sendtransaction) or [wallet](../wallets/index.md). These methods submit the transaction and then return a [TransactionResponse](./transaction-response.md) that allows you to view the result of a transaction at your convenience, leaving the rest of your app processing unblocked.

> However, if you want to send a transaction and wait until it's processed, the SDK also exposes the `sendTransactionAndAwaitStatus` available on a [Provider](../provider/index.md). This method takes a [TransactionRequest](./transaction-request.md) and returns a subsequent result containing it's ID, status and receipts once the node has processed the transaction. Unlike `sendTransaction`, this method maintains the subscription allowing for frictionless transaction processing without any account level abstractions.

Author: https://github.com/FuelLabs/fuels-ts/pull/3541/commits/de4976c519a80c065978c5826c2ba74db4eec5fb

> However, if you want to send a transaction and wait until it's processed, for convenience the SDK also exposes the `sendTransactionAndAwaitStatus` available on a [Provider](../provider/index.md). This method takes a [TransactionRequest](./transaction-request.md) and returns a subsequent result containing it's ID, status and receipts once the node has processed the transaction. Unlike `sendTransaction`, this method maintains the subscription allowing for frictionless transaction processing without any account level abstractions. However, the response is the same as you'd expect from `sendTransaction`.

Reviewer: ```suggestion
However, if you want to send a transaction and wait until it's processed, for convenience the SDK also exposes the `sendTransactionAndAwaitStatus` available on a [Provider](../provider/index.md) which behaves the same as `sendTransaction` but waits until the transaction is processed by the node and then returns the transaction result. 
```

# Transaction Subscriptions

> When submitting transactions via the SDK, usually this is done by calling a method in a [script](../scripts/running-scripts.md) or [contract](../contracts/methods.md#call), or sending a transaction via a [predicate](../predicates/methods.md#sendtransaction) or [wallet](../wallets/index.md). These methods submit the transaction and then returns a [TransactionResponse](./transaction-response.md) that allows you to view the result of a transaction at your convenience, leaving the rest of your app processing unblocked.

Reviewer: ```suggestion
When submitting transactions via the SDK, usually this is done by calling a method in a [script](../scripts/running-scripts.md) or [contract](../contracts/methods.md#call), or sending a transaction via a [predicate](../predicates/methods.md#sendtransaction) or [wallet](../wallets/index.md). These methods submit the transaction and then return a [TransactionResponse](./transaction-response.md) that allows you to view the result of a transaction at your convenience, leaving the rest of your app processing unblocked.
```

> However, if you want to send a transaction and continue subscribing to it's result, the SDK also exposes the `sendTransactionAndAwaitStatus` available on a [Provider](../provider/index.md). This method takes a [TransactionRequest](./transaction-request.md) and returns a minimal transaction result that contains the ID, status and receipts.

Reviewer: ```suggestion
However, if you want to send a transaction and wait until it's processed, the SDK also exposes the `sendTransactionAndAwaitStatus` available on a [Provider](../provider/index.md). This method takes a [TransactionRequest](./transaction-request.md) and returns a minimal transaction result that contains the ID, status and receipts once the node has processed the transaction.
```

This functionality can be used like so:

<<< @./snippets/transaction-subscriptions/transaction-request.ts#main{ts:line-numbers}

Or used with a contract call like so:

<<< @./snippets/transaction-subscriptions/contract-call.ts#main{ts:line-numbers}

> [!NOTE] Note
> This is a blocking call, which means the rest of your app logic could be blocked for several seconds until the transaction is processed. If this is a problem for your app then we recommend using the previous submission methods mentioned in this guide.

> Should you want a more complete transaction summary similar to that provided by a `TransactionResponse`, it can be fetched by it's transaction ID. Check out [this guide](./transaction-response.md) for more information.

Reviewer: ```suggestion
Should you want a more complete transaction summary similar to that provided by a `TransactionResponse`, it can be fetched by its transaction ID. Check out [this guide](./transaction-response.md) for more information.
```
